ros2service/ros2service/verb/echo.py

```diff
@@ -0,0 +1,144 @@
+from rosidl_runtime_py import message_to_csv
+from rosidl_runtime_py import message_to_yaml
+from rosidl_runtime_py.utilities import get_service
+from service_msgs.msg import ServiceEventInfo
+
+DEFAULT_TRUNCATE_LENGTH = 128
+MsgType = TypeVar('MsgType')
+
+
+def unsigned_int(string):
+    try:
+        value = int(string)
+    except ValueError:
+        value = -1
+    if value < 0:
+        raise ArgumentTypeError('value must be non-negative integer')
+    return value
+
+
+class EchoVerb(VerbExtension):
+    """Echo a service."""
+
+    def __init__(self):
+        self.event_type_map = [
+            (v, k) for k, v in ServiceEventInfo._Metaclass_ServiceEventInfo__constants.items()]
+
+    def add_arguments(self, parser, cli_name):
+        arg = parser.add_argument(
+            'service_name',
+            help="Name of the ROS service to echo (e.g. '/add_two_ints')")
+        arg.completer = ServiceNameCompleter(
+            include_hidden_services_key='include_hidden_services')
+        arg = parser.add_argument(
+            'service_type', nargs='?',
+            help="Type of the ROS service (e.g. 'example_interfaces/srv/AddTwoInts')")
+        arg.completer = ServiceTypeCompleter(service_name_key='service_name')
+        parser.add_argument(
+            '--csv', action='store_true', default=False,
+            help=(
+                'Output all recursive fields separated by commas (e.g. for plotting).'
+            ))
+        parser.add_argument(
+            '--full-length', '-f', action='store_true',
+            help='Output all elements for arrays, bytes, and string with a '
+                 "length > '--truncate-length', by default they are truncated "
+                 "after '--truncate-length' elements with '...''")
+        parser.add_argument(
+            '--truncate-length', '-l', type=unsigned_int, default=DEFAULT_TRUNCATE_LENGTH,
+            help='The length to truncate arrays, bytes, and string to '
+                 '(default: %d)' % DEFAULT_TRUNCATE_LENGTH)
+        parser.add_argument(
+            '--no-arr', action='store_true', help="Don't print array fields of messages")
+        parser.add_argument(
+            '--no-str', action='store_true', help="Don't print string fields of messages")
+        parser.add_argument(
+            '--flow-style', action='store_true',
+            help='Print collections in the block style (not available with csv format)')
+
+    def main(self, *, args):
+        if args.service_type is None:
+            with NodeStrategy(args) as node:
+                try:
+                    srv_module = get_service_class(
+                        node, args.service_name, include_hidden_services=True)
+                except (AttributeError, ModuleNotFoundError, ValueError):
+                    raise RuntimeError("The service name '%s' is invalid" % args.service_name)
+        else:
+            try:
+                srv_module = get_service(args.service_type)
```

Comment:
if `ros2 service echo /service1 service/srv/Type2` is issued, it will expects `service1` is `service/srv/Type2`? what if `service1` is `service/srv/Type1`? it could be unexpected behavior?
I think it should check that service type from `service_name` matches with `service_type`?

exception test cases would be ideal.

Reply:
> if ros2 service echo /service1 service/srv/Type2 is issued, it will expects service1 is service/srv/Type2? what if service1 is service/srv/Type1? it could be unexpected behavior?

It currently acts like `ros2 topic echo`.  That is, if you don't give it any type, then it infers the type from the already published topic.  If you give it a type, then it will try to use the type that you gave it.  If the publisher is publishing on that topic already with a different type, you just won't get any data.  `ros2 service echo` is currently setup the same way, where if you give it the "wrong" type, it just won't output any data.

> I think it should check that service type from service_name matches with service_type?

So the main reason we can't do this is that the user may want to start up `ros2 service echo` *before* there are any publishers on the topic.  That way you can see the initial outputs.  In that case, the user *must* provide the type, and we have nothing to compare against.

Overall, I'm not sure we can accomplish both of these things at once.  But I'm open to ideas on how to make this better.

Reply:
that sounds okay, user setting is honored. thanks!